qiskit_experiments/framework/experiment_data.py

```diff
@@ -2364,6 +2378,10 @@ def copy(self, copy_results: bool = True) -> "ExperimentData":
             new_instance._figures = ThreadSafeOrderedDict()
             new_instance.add_figures(self._figures.values())
 
+        with self._artifacts.lock:
+            new_instance._artifacts = ThreadSafeOrderedDict()
+            new_instance.add_artifacts(self._artifacts.values())
```

Comment:
We don't do enough threading for it to matter but `ExperimentData` should probably have a top level lock. Ideally, all the data should be copied within a single lock instead of one per subcomponent.